docs(spring-sdk): add actions as controllers page (#1241)
lightbend/kalix-jvm-sdk: docs/src/modules/spring/pages/actions-as-controller.adoc

= Actions as Controllers

Actions can be used to implement link:https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller[MVC Controllers] by
acting as the external interface of a service, receiving requests, operating over the requests values and forwarding the call
to other components in the same service.

To illustrate how you can use an Action as a Controller, we will build on top of the xref:structuring-your-application.adoc[Value Entity Shopping Cart] example, adding a new Action to the existing shopping cart service.

== Forwarding Commands

The `forward` effect allows us to transform or further validate an incoming request before passing it on to another
component and have the response message directly passed back to the client making the request. The response from the

> forwarded operation must have the same response type as the original command.

Reviewer: I understand this is the same text as the original. I struggled a bit to understand it so here is a suggestion. 
`The `forward` effect allows you to call another
component and have the response message passed back to the method forwarding.  The response from the
forwarded operation must have the same response type as the original method. You can transform or further validate an incoming request before passing it on to another component`

I'm not sure we should have different info than the java docs but at the same time I think is a good opportunity to improve the docs. You may want to check with Renato if you are interested in doing such a change.

Reviewer: > > return the response message to the forwarding method before passing it back to the client making the request
> 
> Don't think this is accurate actually.. because the forward doesn't allow you to transform the response, so effectively the response is passed directly to the client making the request and not to the forwarding method.

Good point. I guess that depends on what before means here. Before in space or in time? (we are doing philosophy now:) Meaning space as in the code, and time as in runtime. I meant in time but I see now it can be interpreted as in the code. In any case, it was a mild suggestion, I guess the original text is just okay the way it is.

Author: Oh I see, makes sense. :) 

Yeah, I couldn't improve it much further I think.. Maybe we can revisit in the future (famous last words).

Reviewer: hehehe okay, let's leave it like it is then.

In this example we accept requests with the same structure as the entity, by receiving a `LineItem`, but add some additional verification of the
request and only conditionally forward the request to the entity if the verification is successful:

.src/main/java/com/example/api/ShoppingCartController.java
[source,java]
----

> include::example$spring-valueentity-shopping-cart/src/main/java/com/example/api/ShoppingCartController.java[tag=forward]

Reviewer: I think it would be nice to explain what is the `ActionCreationContext creationContext` that appears on the constructor.  Especially because `kalixContext` maps to a variable but the `creationContext` doesn't.

Author: I think it should actually work without it, I'll try removing

Reviewer: good idea

----
<1> `KalixClient` is injected on the constructor. It will be used to build calls to the underlining Entity.
<2> Expose the command handler as a `POST` endpoint at specified path.
<3> Check if the added item is carrots.
<4> If it is "carrots" immediately return an error, disallowing adding the item.
<5> For allowed requests, use `kalixClient` to get a `DeferredCall`.

> <6> The `DeferredCall` is used with `effects().forward()` to forward the request to the entity.

Reviewer: Would it be better to give some context of what a `DeferredCall` is?

Reviewer: yeah, I think it helps

NOTE: You might be wondering what the `kalixClient` is about. For now, think of it as a lightweight HTTP client allowing you to reach out to other Kalix services. All details can be found at xref:call-another-service.adoc[] chapter.

== Transform Request and Response to Another Component

The `asyncReply` and `asyncEffect` effects allow us to process and transform a request before calling another component and then also transform the response.

> As an example, let us look at the problem of creating a new entity with an id generated by the service itself.

Reviewer: ```suggestion
As an example, let us look at the problem of creating a new entity with an id generated by another component.
```

Author: for the record, I did this change but also for the java counterpart

<4> Once the call succeeds or fails the `CompletionStage` is completed or failed, we can transform the result from `CompletionStage<Empty>`
to `CompletionStage<Effect<String>>` using `handle`.
<5> On a successful response, create a reply effect passing back the `cartId`.
<6> If the call leads to an error, create an error effect asking the client to retry.
<7> `effects().asyncEffect()` allows us to reply with a `CompletionStage<Effect<String>>`.

The action generates a link:https://en.wikipedia.org/wiki/Universally_unique_identifier[UUID] to use as entity id for the shopping cart. UUIDs are extremely unlikely to lead to the same id
being generated, but to completely guarantee two calls can never be assigned the same shopping cart we
make use of the "boundary of consistency" provided by the entity - the entity will only process a single command at a time
and can safely make decisions based on its state - for example to only allow creation once by storing something in its state
signifying that it has been created.

In this case you mark that the entity has been created using a creation timestamp in the shopping cart state stored on first

> `create` call - when the timestamp has the default value of `0`). If the cart has already been stored with a timestamp it returns an error effect:

Reviewer: ```suggestion
`create` call - when the timestamp has the default value of `0`. If the cart has already been stored with a timestamp it returns an error effect:
```

Also question. Where is this timestamp with the default value of 0? Couldn't find it.

Author: Nice catch!

> Also question. Where is this timestamp with the default value of 0? Couldn't find it.

On the domain model, one of the last lines.. we use that static builder on the `emptyState` override.

> `create` call (when the timestamp has the default value of `0`). If the cart has already been stored with a timestamp we return an error effect:

Reviewer: ```suggestion
`create` call - the timestamp has the default value of `0`. If the cart has already been stored with a timestamp we return an error effect:
```

Author: for the record, I did this change but also for the java counterpart
